File: detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/ViewCommandOperationsReflected.java
package com.wix.detox.reactnative.idlingresources;

import android.util.Log;

import com.facebook.react.uimanager.NativeViewHierarchyManager;
import com.facebook.react.uimanager.UIViewOperationQueue;

import java.util.List;

import org.joor.Reflect;

class ViewCommandOperationsReflected {
    private static final String LOG_TAG = "Detox";
    private final static String CLASS_DISPATCH_COMMAND_OPERATION =
            "com.facebook.react.uimanager.UIViewOperationQueue$DispatchStringCommandOperation";
    private final static String FIELD_MVIEW_COMMAND_OPERATIONS = "mViewCommandOperations";
    private final static String FIELD_NUM_RETRIES = "numRetries";
    private final static String FIELD_MTAG = "mTag";
    private final static String FIELD_MCOMMAND = "mCommand";
    private final static String FIELD_MVIEW_MANAGERS = "mNativeViewHierarchyManager";
    private final static String SET_NATIVE_VALUE = "setNativeValue";
    private final static String REACT_SWITCH = "ReactSwitch";

    // This is a workaround for https://github.com/facebook/react-native/issues/32594
    // uses duck typing heuristics to determine that this is probably the stuck Switch operation and if so, ignores it
    static boolean workaroundForRN66Bug(Object uiOperationQueue) {
        boolean isStuckSwitchOperation = false;

        try {
            Class<?> DispatchStringCommandOperation = Class.forName(CLASS_DISPATCH_COMMAND_OPERATION);
            List<?> mViewCommandOperations = Reflect.on(uiOperationQueue).field(FIELD_MVIEW_COMMAND_OPERATIONS).get();

            if (mViewCommandOperations.size() == 1) {
                Object viewOperation = DispatchStringCommandOperation.cast(mViewCommandOperations.get(0));

                int mTag = Reflect.on(viewOperation).field(FIELD_MTAG).get();
                UIViewOperationQueue uiManagerModule = (UIViewOperationQueue) uiOperationQueue;
                NativeViewHierarchyManager nativeHierarchyManager = Reflect.on(uiManagerModule).field(FIELD_MVIEW_MANAGERS).get();
                String mClassName = nativeHierarchyManager.resolveView(mTag).getClass().getSimpleName();

                int numRetries = Reflect.on(viewOperation).field(FIELD_NUM_RETRIES).get();
                String mCommand = Reflect.on(viewOperation).field(FIELD_MCOMMAND).get();

                boolean isReactSwitch = mClassName.equals(REACT_SWITCH);
                boolean hasOneRetryIncremented = numRetries == 1;
                boolean isSetNativeValueCommand = mCommand.equals(SET_NATIVE_VALUE);

                isStuckSwitchOperation = (isReactSwitch && hasOneRetryIncremented && isSetNativeValueCommand);
            }
        } catch (ClassNotFoundException e) {
            Log.d(LOG_TAG, String.valueOf(e.getCause()));
        }
        return isStuckSwitchOperation;
    }
}

File: detox/android/detox/src/full/java/com/wix/detox/reactnative/idlingresources/ViewCommandOperationsReflected.kt
package com.wix.detox.reactnative.idlingresources

import android.util.Log
import com.facebook.react.uimanager.UIViewOperationQueue
import com.wix.detox.common.DetoxLog.Companion.LOG_TAG
import org.joor.Reflect
import java.lang.ref.WeakReference

private const val FIELD_VIEW_COMMAND_OPERATIONS = "mViewCommandOperations"
private const val FIELD_TAG = "mTag"
private const val FIELD_NUM_RETRIES = "numRetries"
private const val FIELD_COMMAND = "mCommand"

class ViewCommandOperationsReflected(uIViewOperationQueueInstance: UIViewOperationQueue) {
    private val reflectedInstance = Reflect.on(uIViewOperationQueueInstance)

    fun getSize(): Int {
        val viewCommandOperations = viewCommandOperations()
        return viewCommandOperations?.size ?: 0
    }
    fun getTag(): Int {
        return try {
            Reflect.on(firstElement()).field(FIELD_TAG).get<Int>()
        } catch (e: Exception) {
            Log.e(LOG_TAG, "failed to get $FIELD_TAG ", e.cause)
            0
        }
    }
    fun getNumRetries(): Int {
        return try {
            Reflect.on(firstElement()).field(FIELD_NUM_RETRIES).get<Int>()
        } catch (e: Exception) {
            Log.e(LOG_TAG, "failed to get $FIELD_NUM_RETRIES ", e.cause)
            0
        }
    }
    fun getViewCommand(): String {
        return try {
            Reflect.on(firstElement()).field(FIELD_COMMAND).get<String>()
        } catch (e: Exception) {
            Log.e(LOG_TAG, "failed to get $FIELD_COMMAND ", e.cause)
            ""
        }
    }
    fun weakRefFirstElement() = WeakReference(firstElement())
    private fun firstElement() = viewCommandOperations()?.elementAt(0)

    private fun viewCommandOperations(): Collection<Any>? {
        return try {
            reflectedInstance.field(FIELD_VIEW_COMMAND_OPERATIONS).get<Collection<Any>>()
        } catch(e: Exception) {
            Log.e(LOG_TAG, "could not get reflected field mViewCommandOperations ", e.cause)
            null
        }
    }
}
